go/service/user_report.go

```diff
@@ -0,0 +1,145 @@
+// Copyright 2019 Keybase, Inc. All rights reserved. Use of
+// this source code is governed by the included BSD license.
+
+package service
+
+import (
+	"encoding/json"
+	"errors"
+	"fmt"
+	"strings"
+
+	"github.com/keybase/client/go/chat/globals"
+	"github.com/keybase/client/go/libkb"
+	"github.com/keybase/client/go/protocol/chat1"
+	gregor1 "github.com/keybase/client/go/protocol/gregor1"
+	keybase1 "github.com/keybase/client/go/protocol/keybase1"
+	"golang.org/x/net/context"
+)
+
+type convTranscript struct {
+	Messages []convTranscriptMsg `json:"messages"`
+}
+
+type convTranscriptMsgText struct {
+	Body string `json:"body"`
+}
+
+type convTranscriptMsg struct {
+	SenderUsername string                       `json:"senderUsername"`
+	MsgType        string                       `json:"msgType"`
+	Text           *convTranscriptMsgText       `json:"text,omitempty"`
+	Edit           *chat1.MessageEdit           `json:"edit,omitempty"`
+	Delete         *chat1.MessageDelete         `json:"delete,omitempty"`
+	Headline       *chat1.MessageHeadline       `json:"headline,omitempty"`
+	SendPayment    *chat1.MessageSendPayment    `json:"send_payment,omitempty"`
+	RequestPayment *chat1.MessageRequestPayment `json:"request_payment,omitempty"`
+	Unfurl         *chat1.MessageUnfurl         `json:"unfurl,omitempty"`
+}
+
+// When sending a transcript, send the following number of most recent chat
+// messages of selected types.
+const transcriptMessageLimit = 100
+
+func pullTranscript(mctx libkb.MetaContext, chatG *globals.ChatContext, convIDStr string) (res string, err error) {
+	convIDBytes, err := chat1.MakeConvID(convIDStr)
+	if err != nil {
+		return "", err
+	}
+	uidBytes := gregor1.UID(mctx.CurrentUID().ToBytes())
+	chatQuery := &chat1.GetThreadQuery{
+		MarkAsRead: false,
+		MessageTypes: []chat1.MessageType{
+			chat1.MessageType_TEXT,
+			chat1.MessageType_EDIT,
+			chat1.MessageType_DELETE,
+			chat1.MessageType_HEADLINE,
+			chat1.MessageType_SENDPAYMENT,
+			chat1.MessageType_REQUESTPAYMENT,
+			chat1.MessageType_HEADLINE,
+			chat1.MessageType_UNFURL,
+		},
+	}
+	pagination := &chat1.Pagination{
+		Num: transcriptMessageLimit,
+	}
+	threadView, err := chatG.ConvSource.Pull(mctx.Ctx(), convIDBytes, uidBytes, chat1.GetThreadReason_GENERAL,
+		chatQuery, pagination)
+	if err != nil {
+		return "", err
+	}
+	var transcript convTranscript
+	for _, msg := range threadView.Messages {
+		if !msg.IsValid() {
+			continue
+		}
+		mv := msg.Valid()
+		mb := mv.MessageBody
+		var textMsg *convTranscriptMsgText
+		if mb.IsType(chat1.MessageType_TEXT) {
+			textMsg = &convTranscriptMsgText{
+				Body: mb.Text().Body,
+			}
+		}
+		tMsg := convTranscriptMsg{
```

Comment:
Can't you use `json.Marshal` to make this easier?

Reply:
Should we marshal the entire threadView? I wanted to only pass essential fields to save bandwidth and space.

Reply:
Just the `MessageBody` portion of the messages.

Reply:
both comments addressed - good for another look